File: src/demo/repl.js
import readline from 'node:readline/promises';
import { stdin as input, stdout as output } from 'node:process';
import { createEngine, formatState } from './helpers.js';

const engine = createEngine(process.cwd());
const rl = readline.createInterface({ input, output });

console.log('Aleph REPL');
console.log('Type /personas, /switch <id>, /state, or /quit');
console.log(formatState(engine.inspectState()));

try {
  while (true) {
    const line = (await rl.question('\nYou> ')).trim();
    if (!line) continue;

    if (line === '/quit' || line === '/exit') {
      break;
    }

    if (line === '/personas') {
      for (const persona of engine.listPersonas()) {
        console.log(
          `- ${persona.id}: ${persona.name} | specialties=${persona.specialties.join(', ')} | shared=${persona.sharedDomains.join(', ')}`,
        );
      }
      continue;
    }

    if (line === '/state') {
      console.log(formatState(engine.inspectState()));
      continue;
    }

    if (line.startsWith('/switch ')) {
      const targetPersonaId = line.slice('/switch '.length).trim();
      const result = engine.processUserTurn(
        `Foreground handoff requested to ${targetPersonaId}.`,
        { requestedPersonaId: targetPersonaId },
      );
      console.log(`${result.activePersonaName}> ${result.reply}`);
      if (result.switchDecision?.approved) {
        console.log(result.switchDecision.explanation);
      }
      continue;
    }

    const result = engine.processUserTurn(line);
    console.log(`${result.activePersonaName}> ${result.reply}`);
    if (result.switchDecision?.approved) {
      console.log(result.switchDecision.explanation);
    }
  }
} finally {
  rl.close();
  engine.store.close();
}

File: aleph/demo/repl.py
from __future__ import annotations

from pathlib import Path

from aleph.demo.helpers import create_engine, format_state


def main() -> None:
    engine = create_engine(Path.cwd())
    print("Aleph REPL")
    print("Type /clients, /state, /switch <id>, or /quit")
    print(format_state(engine.inspect_state()))
    try:
        while True:
            line = input("\nYou> ").strip()
            if not line:
                continue
            if line in {"/quit", "/exit"}:
                break
            if line == "/clients":
                for client in engine.list_clients():
                    domains = ", ".join(client["declared_capability"].get("domains", []))
                    print(f"- {client['id']}: {client['display_name']} | domains={domains}")
                continue
            if line == "/state":
                print(format_state(engine.inspect_state()))
                continue
            if line.startswith("/switch "):
                target = line[len("/switch ") :].strip()
                events = engine.edge_gateway.submit_text(f"switch to {target}", requested_client_id=target)
            else:
                events = engine.edge_gateway.submit_text(line)

            for event in events:
                payload = event["payload"]
                if event["event_kind"] == "delta":
                    print(f"{event['source']}> {payload['text']}")
                elif event["event_kind"] == "status":
                    print(f"[status] {payload['message']}")
                elif event["event_kind"] == "handoff":
                    print(f"[handoff] {payload['explanation']}")
                elif event["event_kind"] == "tool_event":
                    print(f"[tool] {payload['summary']}")
    finally:
        engine.store.close()
